rust-src/bulletproofs/src/set_membership_proof.rs

```diff
@@ -0,0 +1,702 @@
+//! Implementation of set membership proof along the lines of bulletproofs
+use crate::{inner_product_proof::*, utils::*};
+use crypto_common::*;
+use crypto_common_derive::*;
+use curve_arithmetic::{multiexp, multiexp_table, multiexp_worker_given_table, Curve};
+use ff::Field;
+use pedersen_scheme::*;
+use rand::*;
+use random_oracle::RandomOracle;
+use std::{convert::TryInto, iter::once};
+
+/// Bulletproof style set-membership proof
+#[derive(Clone, Serialize, SerdeBase16Serialize, Debug)]
+#[allow(non_snake_case)]
+pub struct SetMembershipProof<C: Curve> {
+    /// Commitment to the evaluation of the indicator function `I_{v}` on
+    /// the_set
+    A:        C,
+    /// Commitment to the blinding factors in `s_L` and `s_R`
+    S:        C,
+    /// Commitment to the t_1 coefficient of polynomial `t(x)`
+    T_1:      C,
+    /// Commitment to the t_2 coefficient of polynomial `t(x)`
+    T_2:      C,
+    /// Evaluation of t(x) at the challenge point `x`
+    tx:       C::Scalar,
+    /// Blinding factor for the commitment to `tx`
+    tx_tilde: C::Scalar,
+    /// Blinding factor for the commitment to the inner-product arguments
+    e_tilde:  C::Scalar,
+    /// Inner product proof
+    ip_proof: InnerProductProof<C>,
+}
+
+/// Error messages detailing why proof generation failed
+#[derive(Debug, PartialEq, Eq)]
+pub enum ProverError {
+    /// The length of the generator vector `gens` was less than `|the_set|`
+    NotEnoughGenerators,
+    /// Could not find the value `v` in the given set
+    CouldNotFindValueInSet,
+    /// Could not generate inner product proof
+    InnerProductProofFailure,
+    /// Could not invert `y`
+    DivisionError,
+}
+
+/// This function takes a set (as a vector) and a value v as input.
+/// If v in S the function computes bit vectors aL and aR where
+/// aL_i = 1 <=> s_i = v
+/// and a_R is the bit-wise negation of a_L
+/// Note: For multi sets this function only sets the first hit to one, to allow
+/// set membership proofs in multi sets.
+#[allow(non_snake_case)]
+fn a_L_a_R<F: Field>(v: &F, set_vec: &Vec<F>) -> Option<(Vec<F>, Vec<F>)> {
```

Comment:
I think it would be better to take a slice, i.e. `&[F]`,  and then iterate over the slice instead.

Reply:
Why?

Reply:
There is no benefit in passing a &Vec<A> instead of &[A] unless for some reason you care about capacity. The former is a pointer to a pointer, and the latter is just a pointer to a slice. 

See https://rust-lang.github.io/rust-clippy/master/index.html#ptr_arg

Reply:
Thanks!